tsdb/wlog/watcher.go

```diff
@@ -420,9 +437,31 @@ func (w *Watcher) watch(segmentNum int, tail bool) error {
 
 			return nil
 
+		// we haven't read due to a notification in quite some time, try reading anyways
 		case <-readTicker.C:
+			fmt.Println("READ BECAUSE OF TIMEOUT")
 			err = w.readSegment(reader, segmentNum, tail)
+			readTicker.Reset(time.Duration(readTimeout))
+			// Ignore all errors reading to end of segment whilst replaying the WAL.
+			if !tail {
+				if err != nil && errors.Cause(err) != io.EOF {
+					level.Warn(w.logger).Log("msg", "Ignoring error reading to end of segment, may have dropped data", "segment", segmentNum, "err", err)
+				} else if reader.Offset() != size {
+					level.Warn(w.logger).Log("msg", "Expected to have read whole segment, may have dropped data", "segment", segmentNum, "read", reader.Offset(), "size", size)
+				}
+				return nil
+			}
+
+			// Otherwise, when we are tailing, non-EOFs are fatal.
+			if errors.Cause(err) != io.EOF {
+				return err
+			}
 
+		case <-w.readNotify:
```

Comment:
I wonder what would happen if we receive several read notifications while we are already busy processing a read notification. In that case would we want to drain the `readNotify` channel before we begin reading? If that's done, then we would have satisfied all the read requests via just one read.

Reply:
Actually, since this is an unbufferend channel, I guess the Notfy() function would just block in such a situation?

Reply:
I used select with a default case to explicitly not block. 

The theory is that throwing away a notification is not the end of the world. In prometheus' with high scrape load we're probably getting a lot of notifications we don't really need, since for each read by the WAL watcher it reads until EOF. If we were sending a notification "hey, read X bytes" we couldn't get away with throwing these notifications away. In the case of a prometheus with a low scrape load it's very unlikely that we're going to throw away a notification.

This reminds me that adding metrics to track dropping of notifications is a good idea.